src/ppx/geometry.cpp

```diff
@@ -19,6 +19,334 @@
 
 namespace ppx {
 
+// -------------------------------------------------------------------------------------------------
+// VertexDataProcessor
+// -------------------------------------------------------------------------------------------------
+class IVertexDataProcessor
+{
+public:
+    // returns false if the validation fails
+    virtual bool     Validate(Geometry* pGeom)                                        = 0;
+    virtual Result   UpdateVertexBuffer(Geometry* pGeom)                              = 0;
+    virtual uint32_t AppendVertexData(Geometry* pGeom, const TriMeshVertexData& vtx)  = 0;
+    virtual uint32_t AppendVertexData(Geometry* pGeom, const WireMeshVertexData& vtx) = 0;
+    virtual uint32_t GetVertexCount(const Geometry* pGeom)                            = 0;
+
+protected:
+    // Prevent from being deleted explicitly
+    ~IVertexDataProcessor() {}
+
+    // ----------------------------------------
+    // Helper functions to access Geometry data
+    // ----------------------------------------
+
+    // Missing attributes will also result in NOOP.
+    template <typename T>
+    uint32_t AppendDataToVertexBuffer(Geometry* pGeom, uint32_t bufferIndex, const T& data)
+    {
+        if (bufferIndex != PPX_VALUE_IGNORED) {
+            pGeom->mVertexBuffers[bufferIndex].Append(data);
+            return pGeom->mVertexBuffers[bufferIndex].GetElementCount();
+        }
+        return PPX_VALUE_IGNORED;
```

Comment:
Does this match an existing behavior? `PPX_VALUE_IGNORED` is 0xFFFFFFFF. I don't see any checks against. Can you give some details on the intent of the return value if it doesn't match an existing behavior?

Reply:
This is trying to match the behavior of the original function Geometry::AppendPosition(), but since now it is a templated function, the ones for normal, color, tangent, bitangent and texcoord are also returning `ElementCount` (if attribute exist) or `PPX_VALUE_IGNORED` (if the vertex buffer doesn't have that attribute). I could add assert to the return value of position:
```
uint32_t n = AppendDataToVertexBuffer(pGeom, GetPositionBufferIndex(pGeom), vtx.position);
PPX_ASSERT_MSG(PPX_VALUE_IGNORED != n, "position should always available");
```
What do you think?

Reply:
Thanks for the clarification. Had my suspicion that was the case, but wanted to be sure. Neither `PPX_VALUE_IGNORED` or `0` are good return values for these. The intent of the return value is that if it's a valid value, then it could be used for something like calculating indices for a triangle. But both`PPX_VALUE_IGNORED` and `0` can result in incorrect indices. 

Can add a comment to both internal and external functions that `PPX_VALUE_IGNORED` is returned if the bufferIndex is invalid? Maybe also add a range check for `bufferIndex` instead of just against `PPX_VALUE_IGNORED`?

Reply:
change the return value to 0 for invalid case with comments, and added the range check as we discussed.